### PR TITLE
FAI-519: Score Cards: Renaming a DataField does not update Attributes' predicates

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.redhat.com/browse/FAI-519

We were not cascading renames (and validation) into `CompoundPredicates` correctly.

Simple fix.. but I added some tests for areas previously not covered.

### CODE REVIEW

**Comment by manstis** [comment]
"20 queued checks"... nice.

File: packages/pmml-editor/src/editor/paths/PathBuilders.ts [hunk 1]
```diff
@@ -27,7 +27,13 @@ class Builders {
 }
 
 abstract class BaseBuilder {
-  constructor(protected readonly builders: Builders) {}
```
Comment (manstis) [inline packages/pmml-editor/src/editor/paths/PathBuilders.ts:30]: Such sweet, innocuous short hand code.

However it generates `this.builders = builders` after the super constructors are called which overwrites the "cloning" I do [below](https://github.com/kiegroup/kogito-tooling/pull/540/files#diff-94539654eb0b19f9a533482268161d45918f8d5ea2ad64e850443dcd0988e6e7R32). I had to revert to the old skool, boiler plate ridden, approach but at least my code now works the way I expected and I don't need to bang my head on the keyboard any longer to understand WTF was going on :-)

File: packages/pmml-editor/src/editor/paths/PathBuilders.ts [hunk 2]
```diff
@@ -27,7 +27,13 @@ class Builders {
...
+  protected readonly builders: Builders;
+
+  constructor(builders: Builders) {
```
Comment (manstis) [inline packages/pmml-editor/src/editor/paths/PathBuilders.ts:32]: Oh, and BTW, I now do this as when I did [this](https://github.com/kiegroup/kogito-tooling/pull/540/files#diff-551dbb822382201834e50609b0ef69ec355fb1d73873bb245aa9005d655cc068R410) the same collection of builders was being re-used and appended to multiple times leading to crazy voodoo. Each leaf in the path builder can now be safely re-used without fear of the sky crashing down. It does take a little more memory for the cloned array but its content is the same object references i.e. a shallow clone. They are short lived and contain very few (less than 5) object references so I can live with it.

**Comment by tiagobento** [comment]
@manstis ...and 20 successful checks 🚀

**Review by tiagobento — APPROVED** [review]

**Review by kelvah — APPROVED** [review]
LGTM

**Review by ederign — APPROVED** [review]

**Comment by tiagobento** [comment]
Waiting for @sjamboro's review to merge it. cc @manstis

**Comment by manstis** [comment]
@tiagobento FYI @sjamboro is having problems building the PR locally. IDK whether assistance can be given (on Zulip or gChat)? My internet connection is broken at the moment so I was trying to avoid downloading MBs (mobile phone data contract) re-cloning the repo and trying to replicate. Both Jenkins and GH Actions are green.. so it's likely to be a misconfiguration on @sjamboro environment.

File: packages/pmml-editor/src/__tests__/editor/validation/Attributes.test.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    );
+
+    expect(registry.get(asPath("models[0]")).length).toBe(1);
+    assertValidationEntry(
+      "models[0].Characteristics.Characteristic[1].Attribute[2].predicate.fieldName",
+      "cannot be not found"
+    );
+  });
+
+  test("ValidateAttribute::predicates::definedButDoesNotExist::CompoundPredicate", () => {
```
Comment (sjamboro) [inline packages/pmml-editor/src/__tests__/editor/validation/Attributes.test.ts:185]: @manstis I am not able to find test "..predicates::definedAndExist.."
image
How you can see, name of data field was change and validation says "`test` cannot be not found" (I was surprised by double "not") .

Comment (manstis) [reply packages/pmml-editor/src/__tests__/editor/validation/Attributes.test.ts:185]: @sjamboro New commit added fixing text (to "...cannot be found..."). There was multiple occurrences.

I've also added tests for "defined and exists".

I cannot replicate the issue (where field has been renamed but the error `'test' cannot be found` is shown).

Do you have steps (or a file) to reproduce please?

**Comment by manstis** [comment]
@tiagobento Please do not merge yet (I've added a label.. but I know you're trigger happy!)

@sjamboro has reported an issue that does not appear to exist on [http://scorecard.new](http://scorecard.new) so I suspect it maybe a be a regression. If you edit a Characteristic, click away and then try again; then NO characteristics can be edited.

**Comment by manstis** [comment]
OK, it's not a regression (it's on `master/main` too) but let me see if I can squeeze a fix in on here...

**Comment by manstis** [comment]
Issue fixed, this can be merge when approved.

**Comment by sjamboro** [comment]
@tiagobento @manstis approved

**Review by sjamboro — APPROVED** [review]

**Pull request merged**